File: app/src/main/java/com/rolandvitezhu/todocloud/ui/activity/main/fragment/TodoListFragment.java
package com.rolandvitezhu.todocloud.ui.activity.main.fragment;

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.rolandvitezhu.todocloud.R;
import com.rolandvitezhu.todocloud.app.AppController;
import com.rolandvitezhu.todocloud.data.Todo;
import com.rolandvitezhu.todocloud.datastorage.DbLoader;
import com.rolandvitezhu.todocloud.datastorage.asynctask.UpdateViewModelTask;
import com.rolandvitezhu.todocloud.listener.RecyclerViewOnItemTouchListener;
import com.rolandvitezhu.todocloud.receiver.ReminderSetter;
import com.rolandvitezhu.todocloud.ui.activity.main.MainActivity;
import com.rolandvitezhu.todocloud.ui.activity.main.adapter.TodoAdapter;
import com.rolandvitezhu.todocloud.ui.activity.main.dialog.SortTodoListDialog;
import com.rolandvitezhu.todocloud.ui.activity.main.dialogfragment.ConfirmDeleteDialogFragment;
import com.rolandvitezhu.todocloud.ui.activity.main.viewmodel.ListsViewModel;
import com.rolandvitezhu.todocloud.ui.activity.main.viewmodel.PredefinedListsViewModel;
import com.rolandvitezhu.todocloud.ui.activity.main.viewmodel.TodosViewModel;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import javax.inject.Inject;

import androidx.annotation.Nullable;
import androidx.appcompat.view.ActionMode;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;

public class TodoListFragment extends Fragment implements SortTodoListDialog.Presenter {

  @Inject
  DbLoader dbLoader;
  @Inject
  TodoAdapter todoAdapter;

  @BindView(R.id.recyclerview_todolist)
  RecyclerView recyclerView;

  private ActionMode actionMode;

  private TodosViewModel todosViewModel;
  private ListsViewModel listsViewModel;
  private PredefinedListsViewModel predefinedListsViewModel;

  Unbinder unbinder;

  @Override
  public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setHasOptionsMenu(true);

    ((AppController) getActivity().getApplication()).getAppComponent().inject(this);

    todosViewModel = ViewModelProviders.of(getActivity()).get(TodosViewModel.class);
    listsViewModel = ViewModelProviders.of(getActivity()).get(ListsViewModel.class);
    predefinedListsViewModel = ViewModelProviders.of(getActivity()).get(PredefinedListsViewModel.class);

    todosViewModel.getTodos().observe(
        this,
        new Observer<ArrayList<Todo>>() {

          @Override
          public void onChanged(@Nullable ArrayList<Todo> todos) {
            todoAdapter.update(todos);
            todoAdapter.notifyDataSetChanged();
          }
        }
    );

    updateTodosViewModel();
  }

  @Nullable
  @Override
  public View onCreateView(LayoutInflater inflater,
                           @Nullable ViewGroup container,
                           @Nullable Bundle savedInstanceState
  ) {
    View view = inflater.inflate(R.layout.fragment_todolist, container, false);
    unbinder = ButterKnife.bind(this, view);

    prepareRecyclerView(view);
    applyClickEvents();
    applySwipeToDismissAndDragToReorder();

    return view;
  }

  @Override
  public void onResume() {
    super.onResume();
    setActionBarTitle();
  }

  @Override
  public void onDestroyView() {
    super.onDestroyView();
    unbinder.unbind();
  }

  private void prepareRecyclerView(View view) {
    RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(
        getContext().getApplicationContext()
    );
    recyclerView.setLayoutManager(layoutManager);
    recyclerView.setAdapter(todoAdapter);
  }

  private void applyClickEvents() {
    recyclerView.addOnItemTouchListener(new RecyclerViewOnItemTouchListener(
        getContext().getApplicationContext(),
        recyclerView,
        new RecyclerViewOnItemTouchListener.OnClickListener() {

          @Override
          public void onClick(View childView, int childViewAdapterPosition) {
            if (!isActionMode()) {
              openModifyTodoFragment(childViewAdapterPosition);
            } else {
              todoAdapter.toggleSelection(childViewAdapterPosition);

              if (areSelectedItems()) {
                actionMode.invalidate();
              } else {
                actionMode.finish();
              }
            }
          }

          @Override
          public void onLongClick(View childView, int childViewAdapterPosition) {
            if (!isActionMode()) {
              ((MainActivity)TodoListFragment.this.getActivity()).onStartActionMode(callback);
              todoAdapter.toggleSelection(childViewAdapterPosition);
              actionMode.invalidate();
            }
          }

        }
        )
    );
  }

  private void applySwipeToDismissAndDragToReorder() {
    ItemTouchHelper.SimpleCallback simpleItemTouchCallback = new ItemTouchHelper.SimpleCallback(
        0, ItemTouchHelper.START
    ) {

      @Override
      public boolean onMove(
          RecyclerView recyclerView,
          RecyclerView.ViewHolder viewHolder,
          RecyclerView.ViewHolder target
      ) {
        List<Todo> todos = todoAdapter.getTodos();
        int fromPosition = viewHolder.getAdapterPosition();
        int toPosition = target.getAdapterPosition();
        if (fromPosition < toPosition) {
          for (int i = fromPosition; i < toPosition; i++) {
            swapItems(todos, i, i + 1);
          }
        } else {
          for (int i = fromPosition; i > toPosition; i--) {
            swapItems(todos, i, i - 1);
          }
        }
        todoAdapter.notifyItemMoved(viewHolder.getAdapterPosition(), target.getAdapterPosition());

//        if (target.getAdapterPosition() < viewHolder.getAdapterPosition()) {
//          todos.get(target.getAdapterPosition())
//        }

//        // TODO: implement reoder/persist reorder
//        // 0. DONE  Persist, when change the item's position value - set the item's dirty value to true
//        // 1. Order in an AsyncTask
//        // 2. The items, that changed it's position value should have a position value that is
//        // halfway between the previous and the next item's position values.
//        // 3. Implement more efficient order algorithm
//        boolean didOrderLastTime = false;
//        do {
//          didOrderLastTime = false;
//          for (int i = 0; i < todos.size()-1; i++) {
//            // Order, if should
//            if (todos.get(i).getPosition() >= todos.get(i+1).getPosition()) {
//              todos.get(i).setPosition(todos.get(i).getPosition() - 51);
//              todos.get(i).setDirty(true);
//              dbLoader.updateTodo(todos.get(i));
//              didOrderLastTime = true;
//            }
//          }
//        } while (didOrderLastTime); // If it's ordered correctly, then the ordering process is done

//        AsyncTask persistReorderAsyncTask = new AsyncTask() {
//          @Override
//          protected Object doInBackground(Object[] objects) {
//
//
//            return null;
//          }
//        }.execute();

        return true;
      }

      @Override
      public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
        Todo swipedTodo = getSwipedTodo(viewHolder);
        int swipedTodoAdapterPosition = viewHolder.getAdapterPosition();
        openConfirmDeleteTodosDialog(swipedTodo, swipedTodoAdapterPosition);
      }

      @Override
      public int getMovementFlags(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
        int dragFlags;
        int swipeFlags;
        if (AppController.isActionMode()) {
          dragFlags = ItemTouchHelper.UP | ItemTouchHelper.DOWN;
          swipeFlags = 0;
        }
        else {
          dragFlags = 0;
          swipeFlags = ItemTouchHelper.START;
        }
        return makeMovementFlags(dragFlags, swipeFlags);
      }

      @Override
      public boolean isLongPressDragEnabled() {
        return false;
      }
    };
    ItemTouchHelper itemTouchHelper = new ItemTouchHelper(simpleItemTouchCallback);
    itemTouchHelper.attachToRecyclerView(recyclerView);
    todoAdapter.setItemTouchHelper(itemTouchHelper);
  }

  private void swapItems(List<Todo> todos, int fromPosition, int toPosition) {
    // Change todo position values
    Todo todoFrom = todos.get(fromPosition);
    Todo todoTo = todos.get(toPosition);
    int tempTodoToPosition = todoFrom.getPosition();
    todoFrom.setPosition(todoTo.getPosition());
    todoTo.setPosition(tempTodoToPosition);
    todoFrom.setDirty(true);
    todoTo.setDirty(true);
    dbLoader.updateTodo(todoFrom);
    dbLoader.updateTodo(todoTo);
    dbLoader.fixTodoPositions();
    Collections.swap(todos, fromPosition, toPosition);
  }

  private Todo getSwipedTodo(RecyclerView.ViewHolder viewHolder) {
    int swipedTodoAdapterPosition = viewHolder.getAdapterPosition();
    return todoAdapter.getTodo(swipedTodoAdapterPosition);
  }

  private boolean areSelectedItems() {
    return todoAdapter.getSelectedItemCount() > 0;
  }

  private boolean isActionMode() {
    return actionMode != null;
  }

  private void openModifyTodoFragment(int childViewAdapterPosition) {
    Todo todo = todoAdapter.getTodo(childViewAdapterPosition);
    todosViewModel.setTodo(todo);

    ((MainActivity)getActivity()).openModifyTodoFragment(this);
  }

  @Override
  public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
    inflater.inflate(R.menu.fragment_todolist, menu);
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    int menuItemId = item.getItemId();

    switch (menuItemId) {
      case R.id.menuitem_todolist_sort:
        new SortTodoListDialog(getActivity(), this);
        break;
    }

    return super.onOptionsItemSelected(item);
  }

  private void setActionBarTitle() {
    String actionBarTitle = "";

    if (!todosViewModel.isPredefinedList()) // List
      actionBarTitle = listsViewModel.getList().getTitle();
    else // PredefinedList
      actionBarTitle = predefinedListsViewModel.getPredefinedList().getTitle();

    ((MainActivity)getActivity()).onSetActionBarTitle(actionBarTitle);
  }

  private ActionMode.Callback callback = new ActionMode.Callback() {

    @Override
    public boolean onCreateActionMode(ActionMode mode, Menu menu) {
      setActionMode(mode);
      mode.getMenuInflater().inflate(R.menu.layout_appbar_todolist, menu);

      return true;
    }

    @Override
    public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
      String title = prepareTitle();
      actionMode.setTitle(title);
      todoAdapter.notifyDataSetChanged();

      return true;
    }

    @Override
    public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
      int actionItemId = item.getItemId();

      switch (actionItemId) {
        case R.id.menuitem_layoutappbartodolist_delete:
          openConfirmDeleteTodosDialog();
          break;
      }

      return true;
    }

    @Override
    public void onDestroyActionMode(ActionMode mode) {
      todoAdapter.clearSelection();
      setActionMode(null);
      todoAdapter.notifyDataSetChanged();
    }

    private String prepareTitle() {
      int selectedItemCount = todoAdapter.getSelectedItemCount();
      String title = selectedItemCount + " " + getString(R.string.all_selected);
      return title;
    }

  };

  private void setActionMode(ActionMode actionMode) {
    this.actionMode = actionMode;
    AppController.setActionMode(actionMode);
  }

  private void openConfirmDeleteTodosDialog() {
    ArrayList<Todo> selectedTodos = todoAdapter.getSelectedTodos();
    Bundle arguments = new Bundle();
    arguments.putString("itemType", "todo");
    arguments.putParcelableArrayList("itemsToDelete", selectedTodos);
    openConfirmDeleteDialogFragment(arguments);
  }

  private void openConfirmDeleteTodosDialog(Todo swipedTodo, int swipedTodoAdapterPosition) {
    ArrayList<Todo> selectedTodos = new ArrayList<>();
    selectedTodos.add(swipedTodo);
    Bundle arguments = new Bundle();
    arguments.putString("itemType", "todo");
    arguments.putParcelableArrayList("itemsToDelete", selectedTodos);
    openConfirmDeleteDialogFragment(arguments, swipedTodoAdapterPosition);
  }

  private void openConfirmDeleteDialogFragment(Bundle arguments) {
    ConfirmDeleteDialogFragment confirmDeleteDialogFragment = new ConfirmDeleteDialogFragment();
    confirmDeleteDialogFragment.setTargetFragment(this, 0);
    confirmDeleteDialogFragment.setArguments(arguments);
    confirmDeleteDialogFragment.show(getFragmentManager(), "ConfirmDeleteDialogFragment");
  }

  private void openConfirmDeleteDialogFragment(
      Bundle arguments, final int swipedTodoAdapterPosition
  ) {
    ConfirmDeleteDialogFragment confirmDeleteDialogFragment = new ConfirmDeleteDialogFragment();
    confirmDeleteDialogFragment.setTargetFragment(this, 0);
    confirmDeleteDialogFragment.setArguments(arguments);
    confirmDeleteDialogFragment.show(getFragmentManager(), "ConfirmDeleteDialogFragment");
    applyDismissEvents(swipedTodoAdapterPosition, confirmDeleteDialogFragment);
  }

  private void applyDismissEvents(
      final int swipedTodoAdapterPosition, ConfirmDeleteDialogFragment confirmDeleteDialogFragment
  ) {
    getFragmentManager().executePendingTransactions();
    Dialog confirmDeleteDialogFragmentDialog = confirmDeleteDialogFragment.getDialog();
    confirmDeleteDialogFragmentDialog.setOnDismissListener(
        new DialogInterface.OnDismissListener() {

          @Override
          public void onDismiss(DialogInterface dialog) {
            todoAdapter.notifyItemChanged(swipedTodoAdapterPosition);
          }

        }
    );
  }

  private void updateTodosViewModel() {
    UpdateViewModelTask updateViewModelTask = new UpdateViewModelTask(todosViewModel, getActivity());
    updateViewModelTask.execute();
  }

  public void onSoftDelete(String onlineId, String itemType) {
    Todo todoToSoftDelete = dbLoader.getTodo(onlineId);
    dbLoader.softDeleteTodo(todoToSoftDelete);
    updateTodosViewModel();
    ReminderSetter.cancelReminderService(todoToSoftDelete);
    if (actionMode != null) {
      actionMode.finish();
    }
  }

  public void onSoftDelete(ArrayList itemsToDelete, String itemType) {
    ArrayList<Todo> todosToSoftDelete = itemsToDelete;
    for (Todo todoToSoftDelete:todosToSoftDelete) {
      dbLoader.softDeleteTodo(todoToSoftDelete);
      ReminderSetter.cancelReminderService(todoToSoftDelete);
    }
    updateTodosViewModel();
    if (actionMode != null) {
      actionMode.finish();
    }
  }

  @Override
  public void onSortByDueDatePushed() {
    new SortAsyncTask(todoAdapter).execute(SortAsyncTask.SORT_BY_DUE_DATE);
  }

  @Override
  public void onSortByPriorityPushed() {
    new SortAsyncTask(todoAdapter).execute(SortAsyncTask.SORT_BY_PRIORITY);
  }

  @OnClick(R.id.floatingactionbutton_todolist)
  public void onFABClick(View view) {
    if (isActionMode()) actionMode.finish();
    ((MainActivity)TodoListFragment.this.getActivity()).
        onOpenCreateTodoFragment(TodoListFragment.this);
  }

  private static class SortAsyncTask extends AsyncTask<Integer, Long, Void> {

    public static final int SORT_BY_DUE_DATE = 1001;
    public static final int SORT_BY_PRIORITY = 1002;
    
    private WeakReference<TodoAdapter> todoAdapterWeakReference;

    SortAsyncTask(TodoAdapter context) {
      todoAdapterWeakReference = new WeakReference<>(context);
    }

    @Override
    protected Void doInBackground(Integer... params) {
      switch (params[0]) {
        case SORT_BY_DUE_DATE:
          todoAdapterWeakReference.get().sortByDueDate();
          break;
        case SORT_BY_PRIORITY:
          todoAdapterWeakReference.get().sortByPriority();
          break;
      }

      return null;
    }

    @Override
    protected void onPostExecute(Void aVoid) {
      super.onPostExecute(aVoid);
      TodoAdapter todoAdapter = todoAdapterWeakReference.get();
      if (todoAdapter == null) return;

      todoAdapter.notifyDataSetChanged();
    }
  }

}

File: app/src/main/java/com/rolandvitezhu/todocloud/ui/activity/main/fragment/TodoListFragment.kt
package com.rolandvitezhu.todocloud.ui.activity.main.fragment

import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.view.*
import androidx.appcompat.view.ActionMode
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import com.rolandvitezhu.todocloud.R
import com.rolandvitezhu.todocloud.app.AppController
import com.rolandvitezhu.todocloud.app.AppController.Companion.instance
import com.rolandvitezhu.todocloud.app.AppController.Companion.isDraggingEnabled
import com.rolandvitezhu.todocloud.data.Todo
import com.rolandvitezhu.todocloud.database.TodoCloudDatabaseDao
import com.rolandvitezhu.todocloud.databinding.FragmentTodolistBinding
import com.rolandvitezhu.todocloud.ui.activity.main.MainActivity
import com.rolandvitezhu.todocloud.ui.activity.main.adapter.TodoAdapter
import com.rolandvitezhu.todocloud.ui.activity.main.dialogfragment.ConfirmDeleteDialogFragment
import com.rolandvitezhu.todocloud.ui.activity.main.dialogfragment.SortTodoListDialogFragment
import com.rolandvitezhu.todocloud.ui.activity.main.viewmodel.ListsViewModel
import com.rolandvitezhu.todocloud.ui.activity.main.viewmodel.PredefinedListsViewModel
import com.rolandvitezhu.todocloud.ui.activity.main.viewmodel.TodosViewModel
import kotlinx.android.synthetic.main.layout_recyclerview_todolist.view.*
import kotlinx.coroutines.launch
import java.util.*
import javax.inject.Inject

class TodoListFragment : Fragment(), DialogInterface.OnDismissListener {

    @Inject
    lateinit var todoCloudDatabaseDao: TodoCloudDatabaseDao
    @Inject
    lateinit var todoAdapter: TodoAdapter

    var actionMode: ActionMode? = null

    private val todosViewModel by lazy {
        activity?.let { ViewModelProvider(it).get(TodosViewModel::class.java) }
    }
    private val listsViewModel by lazy {
        activity?.let { ViewModelProvider(it).get(ListsViewModel::class.java) }
    }
    private val predefinedListsViewModel by lazy {
        activity?.let { ViewModelProvider(it).get(PredefinedListsViewModel::class.java) }
    }

    private var swipedTodoAdapterPosition: Int? = null

    override fun onAttach(context: Context) {
        super.onAttach(context)
        Objects.requireNonNull(instance)?.appComponent?.fragmentComponent()?.create()?.inject(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)

        todosViewModel?.todos?.observe(
                this,
                Observer { todos ->
                    todos?.let { todoAdapter.update(it.toMutableList()) }
                    todoAdapter.notifyDataSetChanged()
                }
        )
    }

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?
    ): View? {
        val fragmentTodoListBinding: FragmentTodolistBinding =
                FragmentTodolistBinding.inflate(inflater, container, false)
        val view: View = fragmentTodoListBinding.root

        prepareRecyclerView()
        applySwipeToDismissAndDragToReorder(view)

        fragmentTodoListBinding.lifecycleOwner = this
        fragmentTodoListBinding.todoListFragment = this
        fragmentTodoListBinding.executePendingBindings()

        return view
    }

    override fun onResume() {
        super.onResume()
        setActionBarTitle()
    }

    private fun prepareRecyclerView() {
        isDraggingEnabled = true
    }

    private fun applySwipeToDismissAndDragToReorder(view: View) {
        val simpleItemTouchCallback: ItemTouchHelper.SimpleCallback =
                object : ItemTouchHelper.SimpleCallback(
                0, ItemTouchHelper.START
        ) {
            override fun onMove(
                    recyclerView: RecyclerView,
                    viewHolder: RecyclerView.ViewHolder,
                    target: RecyclerView.ViewHolder
            ): Boolean {
                val todos: List<Todo?> = todoAdapter.getTodos()
                val fromPosition = viewHolder.adapterPosition
                val toPosition = target.adapterPosition

                // We swap the position of the todos on the list by moving them. We swap the
                // position values of the moved todos.
                if (fromPosition < toPosition) {
                    for (i in fromPosition until toPosition) {
                        swapItems(todos, i, i + 1)
                    }
                } else {
                    for (i in fromPosition downTo toPosition + 1) {
                        swapItems(todos, i, i - 1)
                    }
                }

                todoAdapter.notifyItemMoved(viewHolder.adapterPosition, target.adapterPosition)

                return true
            }

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                val swipedTodo = getSwipedTodo(viewHolder)
                swipedTodoAdapterPosition = viewHolder.adapterPosition
                swipedTodoAdapterPosition?.let { openConfirmDeleteTodosDialog(swipedTodo, it) }
            }

            override fun getMovementFlags(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder): Int {
                val dragFlags: Int
                val swipeFlags: Int
                if (AppController.isActionMode()) {
                    dragFlags = ItemTouchHelper.UP or ItemTouchHelper.DOWN
                    swipeFlags = 0
                } else {
                    dragFlags = 0
                    swipeFlags = ItemTouchHelper.START
                }

                return ItemTouchHelper.Callback.makeMovementFlags(dragFlags, swipeFlags)
            }

            override fun isLongPressDragEnabled(): Boolean {
                return false
            }
        }

        val itemTouchHelper = ItemTouchHelper(simpleItemTouchCallback)
        itemTouchHelper.attachToRecyclerView(view.recyclerview_todolist)
        todoAdapter.itemTouchHelper = itemTouchHelper
    }

    /**
     * Change the todo position values.
     */
    private fun swapItems(todos: List<Todo?>, fromPosition: Int, toPosition: Int) {
        val todoFrom = todos[fromPosition]
        val todoTo = todos[toPosition]
        if (todoFrom != null && todoTo != null) {
            val tempTodoToPosition = todoFrom.position

            todoFrom.position = todoTo.position
            todoTo.position = tempTodoToPosition
            todoFrom.dirty = true
            todoTo.dirty = true

            Collections.swap(todos, fromPosition, toPosition)

            lifecycleScope.launch {
                persistSwappedItems(todoFrom, todoTo)
            }
        }
    }

    /**
     * Update the position value of the todo items in the database as well and fix the duplications
     * of the positions if there are.
     */
    private suspend fun persistSwappedItems(todoFrom: Todo?, todoTo: Todo?) {
        todoFrom?.let{ todoCloudDatabaseDao.updateTodo(it) }
        todoTo?.let{ todoCloudDatabaseDao.updateTodo(it) }
        todoCloudDatabaseDao.fixTodoPositions()
    }

    private fun getSwipedTodo(viewHolder: RecyclerView.ViewHolder): Todo {
        val swipedTodoAdapterPosition = viewHolder.adapterPosition
        return todoAdapter.getTodo(swipedTodoAdapterPosition)
    }

    fun areSelectedItems(): Boolean {
        return todoAdapter.selectedItemCount > 0
    }

    fun isActionMode() = actionMode != null

    fun openModifyTodoFragment(childViewAdapterPosition: Int) {
        todosViewModel?.initToModifyTodo(todoAdapter.getTodo(childViewAdapterPosition))
        (activity as MainActivity?)?.openModifyTodoFragment(this, null)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.fragment_todolist, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menuitem_todolist_sort -> openSortTodoListDialogFragment()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun openSortTodoListDialogFragment() {
        val sortTodoListDialogFragment = SortTodoListDialogFragment()
        sortTodoListDialogFragment.setTargetFragment(this, 0)
        sortTodoListDialogFragment.show(parentFragmentManager, "SortTodoListDialogFragment")
    }

    private fun setActionBarTitle() {
        (activity as MainActivity?)?.onSetActionBarTitle(todosViewModel?.todosTitle)
    }

    val callback: ActionMode.Callback = object : ActionMode.Callback {
        override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
            setTodoListActionMode(mode)
            mode.menuInflater.inflate(R.menu.layout_appbar_todolist, menu)
            return true
        }

        override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
            val title = prepareTitle()
            actionMode?.title = title
            todoAdapter.notifyDataSetChanged()
            return true
        }

        override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
            when (item.itemId) {
                R.id.menuitem_layoutappbartodolist_delete -> openConfirmDeleteTodosDialog()
            }
            return true
        }

        override fun onDestroyActionMode(mode: ActionMode) {
            todoAdapter.clearSelection()
            setTodoListActionMode(null)
            todoAdapter.notifyDataSetChanged()
        }

        private fun prepareTitle(): String {
            val selectedItemCount = todoAdapter.selectedItemCount
            return selectedItemCount.toString() + " " + getString(R.string.all_selected)
        }
    }

    private fun setTodoListActionMode(actionMode: ActionMode?) {
        this.actionMode = actionMode
        AppController.setActionMode(actionMode)
    }

    private fun openConfirmDeleteTodosDialog() {
        val selectedTodos = todoAdapter.selectedTodos
        val arguments = Bundle()
        arguments.putString("itemType", "todo")
        arguments.putParcelableArrayList("itemsToDelete", selectedTodos)
        openConfirmDeleteDialogFragment(arguments)
    }

    private fun openConfirmDeleteTodosDialog(swipedTodo: Todo, swipedTodoAdapterPosition: Int) {
        val selectedTodos = ArrayList<Todo>()
        selectedTodos.add(swipedTodo)
        val arguments = Bundle()
        arguments.putString("itemType", "todo")
        arguments.putParcelableArrayList("itemsToDelete", selectedTodos)
        openConfirmDeleteDialogFragment(arguments)
    }

    private fun openConfirmDeleteDialogFragment(arguments: Bundle) {
        val confirmDeleteDialogFragment = ConfirmDeleteDialogFragment()
        confirmDeleteDialogFragment.setTargetFragment(this, 0)
        confirmDeleteDialogFragment.arguments = arguments
        confirmDeleteDialogFragment.show(parentFragmentManager, "ConfirmDeleteDialogFragment")
    }

    fun onFABClick() {
        actionMode?.finish()
        (this@TodoListFragment.activity as MainActivity?)?.
        onOpenCreateTodoFragment(this@TodoListFragment)
    }

    override fun onDismiss(dialog: DialogInterface?) {
        if (swipedTodoAdapterPosition != null)
            todoAdapter.notifyItemChanged(swipedTodoAdapterPosition!!)
        else
            todoAdapter.notifyDataSetChanged()
    }

    /**
     * Finish the action mode, if the fragment is in action mode.
     */
    fun finishActionMode() {
        actionMode?.finish()
    }
}